File: python/satellite/satellite.py
def tree_from_traversals(preorder, inorder):
    if len(preorder) != len(inorder):
        raise ValueError('Traversal lengths should be same!')
    if len(preorder) != len(set(preorder)) or len(inorder) != len(set(inorder)):
        raise ValueError('Items in one traversal should be distinct!')

    return build(preorder, 0, len(preorder) - 1, inorder, 0, len(inorder) - 1)


def build(preorder, pre_begin_index, pre_end_index, inorder, in_begin_index, in_end_index):
    if pre_begin_index > pre_end_index:
        return {}

    in_index = in_begin_index
    while inorder[in_index] != preorder[pre_begin_index]:
        in_index += 1
        if in_index == len(inorder):
            raise ValueError('Inconsistent traversals!')

    return {
        'v': preorder[pre_begin_index],
        'l': build(preorder, pre_begin_index + 1, pre_begin_index + (in_index - in_begin_index), inorder, in_begin_index, in_index - 1),
        'r': build(preorder, pre_begin_index + (in_index - in_begin_index) + 1, pre_end_index, inorder, in_index + 1, in_end_index)
    }

File: java/satellite/src/main/java/Satellite.java
import java.util.HashSet;
import java.util.List;

public class Satellite {
    public Tree treeFromTraversals(List<Character> preorder, List<Character> inorder) {
        if (preorder.size() != inorder.size()) {
            throw new IllegalArgumentException("traversals must have the same length");
        }
        if (!new HashSet<>(preorder).equals(new HashSet<>(inorder))) {
            throw new IllegalArgumentException("traversals must have the same elements");
        }
        if (preorder.stream().distinct().count() != preorder.size()
                || inorder.stream().distinct().count() != inorder.size()) {
            throw new IllegalArgumentException("traversals must contain unique items");
        }

        return new Tree(build(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1));
    }

    Node build(List<Character> preorder, int preBeginIndex, int preEndIndex, List<Character> inorder, int inBeginIndex,
            int inEndIndex) {
        if (preBeginIndex > preEndIndex) {
            return null;
        }

        int inIndex = inBeginIndex;
        while (!inorder.get(inIndex).equals(preorder.get(preBeginIndex))) {
            ++inIndex;
        }

        return new Node(preorder.get(preBeginIndex),
                build(preorder, preBeginIndex + 1, preBeginIndex + (inIndex - inBeginIndex), inorder, inBeginIndex,
                        inIndex - 1),
                build(preorder, preBeginIndex + (inIndex - inBeginIndex) + 1, preEndIndex, inorder, inIndex + 1,
                        inEndIndex));
    }
}
